modules/miner/blockmanager.go

```diff
@@ -69,21 +69,53 @@ func (m *Miner) HeaderForWork() (types.BlockHeader, types.Target) {
 	lockID := m.mu.Lock()
 	defer m.mu.Unlock(lockID)
 
-	// Grab a block for work.
-	block, target := m.blockForWork()
+	var header types.BlockHeader
+	var randTxn types.Transaction
+	var block *types.Block
 
-	// Save a mapping between the block and its header, replacing the block
-	// that was stored 'headerForWorkMemory' requests ago.
+	if m.memProgress%headersPerBlockMemory == 0 {
+		// Grab a new block
+		block = new(types.Block)
+		*block, _ = m.blockForWork()
```

Comment:
I suggest not using a pointer.
```go
var block types.Block
if m.memProgress%headersPerBlockMemory == 0 {
    block, _ = m.blockForWork()
} else {
    block = *m.blockMem[m.headerMem[m.memProgress-1]]
}
```
perhaps you could even make blockMem a `[]types.Block` instead of a `[]*types.Block`